src/box/box.cc

```diff
@@ -681,6 +681,234 @@ box_check_uri(const char *source, const char *option_name)
 	return 0;
 }
 
+static int
+box_check_uri_option_val(const char *source, const char *option_name)
+{
+	if (source == NULL)
+		return 0;
+
+
+	char **option_val_array = NULL, *source_copy = NULL;
+	unsigned size = 0;
+	int rc = 0;
+
+	source_copy = strdup(source);
+	if (source_copy == NULL) {
+		diag_set(OutOfMemory, strlen(source) + 1,
+			 "strdup", "source");
+		rc = -1;
+		goto cleanup;
+	}
+	option_val_array = str_split(source_copy, ";", &size, false);
+	if (option_val_array == NULL) {
+		diag_set(OutOfMemory, size * sizeof(char *),
+			 "str_split", "option_val_array");
+		rc = -1;
+		goto cleanup;
+	}
+
+	for (unsigned i = 0; i < size; i++) {
+		if (!strcmp(option_name, "transport")) {
+			if (strcmp(option_val_array[i], "plain")) {
+				diag_set(ClientError, ER_CFG, option_name,
+					 "expected plain value for uri "
+					 "transport option");
+				rc = -1;
+				break;
+			}
+		} else {
+			diag_set(ClientError, ER_CFG, option_name,
+				 "unexpected option");
+				rc = -1;
+				break;
+		}
+	}
+
+cleanup:
+	if (option_val_array != NULL)
+		free(option_val_array);
+	if (source_copy != NULL)
+		free(source_copy);
+	return rc;
+}
+
+static int
+box_check_uri_option(const char *source, const char *option_name)
+{
+	if (source == NULL)
+		return 0;
+
+	char *source_copy = strdup(source);
+	if (source_copy == NULL) {
+		diag_set(OutOfMemory, strlen(source) + 1,
+			 "strdup", "source");
+		return -1;
+	}
+	int rc = 0;
+	char *opt_name = strtok(source_copy, "=");
+	char *opt_val = strtok(NULL, "=");
+	if (opt_name == NULL || opt_name != source_copy) {
+		diag_set(ClientError, ER_CFG, option_name,
+			 "expected uri?uri_option=uri_option_values");
+		rc = -1;
+		goto finish;
+	}
+	if (opt_val == NULL || strtok(NULL, "=") != NULL) {
+		diag_set(ClientError, ER_CFG, option_name,
+			 "expected uri?uri_option=uri_option_values");
+		rc = -1;
+		goto finish;
+	}
+	rc = box_check_uri_option_val(opt_val, opt_name);
+finish:
+	free(source_copy);
+	return rc;
+}
+
+static int
+box_check_uri_options(const char *source, const char *option_name)
+{
+	if (source == NULL)
+		return 0;
+
+	char **options_array = NULL, *source_copy = NULL;
+	unsigned size = 0;
+	int rc = 0;
+
+	source_copy = strdup(source);
+	if (source_copy == NULL) {
+		diag_set(OutOfMemory, strlen(source) + 1,
+			 "strdup", "source");
+		rc = -1;
+		goto cleanup;
+	}
+	options_array = str_split(source_copy, "&", &size, false);
+	if (options_array == NULL) {
+		if (size != 0) {
+			diag_set(OutOfMemory, size * sizeof(char *),
+				 "str_split", "options_array");
+		} else {
+			diag_set(ClientError, ER_CFG, option_name,
+				 "uri_options can not be empty "
```

Comment:
This error repeats four times so I'd consider to make it separate ER_ error. Alternatively, you can add helper which would set this error (alongside with checking str).